nx/lib/nx/defn/grad.ex

```diff
@@ -213,6 +265,31 @@ defmodule Nx.Defn.Grad do
     end
   end
 
+  defp compute_arg_vectorized_names(%{vectorized_axes: vectorized_axes}, []),
+    do: Keyword.keys(vectorized_axes)
+
+  defp compute_arg_vectorized_names(
+         %{vectorized_axes: vectorized_axes, names: names},
+         parent_names
+       ) do
+    reversed_inner_axes =
+      Enum.reduce(names, [], fn name, acc ->
+        if name in parent_names do
+          [name | acc]
+        else
+          acc
+        end
+      end)
+
+    Keyword.keys(vectorized_axes) ++ Enum.reverse(reversed_inner_axes)
```

Comment:
This is a Enum.filter, no? 

```suggestion
    Keyword.keys(vectorized_axes) ++ Enum.filter(names, & &1 in parent_names)
```

Reply:
Yup. The loop had to iterate over multiple data structures before